engine.py

```diff
@@ -193,20 +203,80 @@ def undo_move(self, look_ahead_mode = False):
 		"""
 		if self.move_log:
 			last_move = self.move_log.pop()
+
 			self.board[last_move.start_row][last_move.start_col] = last_move.piece_moved
 			self.board[last_move.end_row][last_move.end_col] = last_move.piece_captured
 			self.light_to_move = not self.light_to_move
+			###
+			#Undoing kings movement
+			if last_move.piece_moved == "kl": #  light king pices moved undo
+				self.light_king_pos = (last_move.start_row, last_move.start_col) # update tuple by undoing move
+			elif last_move.piece_moved == "kd": #  Dark king pices moved undo
+				self.dark_king_pos = (last_move.start_row, last_move.start_col) # update tuple by undoing move
+			self.undo_text = "undoing -> {}".format(last_move.get_chess_notation()) # an undoing statement to be printed to show undo done
+
 
-			print("undoing ->", last_move.get_chess_notation())
 		else:
 			print("All undone!")
 
 
-	def get_valid_moves(self):
-		return self.get_possible_moves()
+
+	def get_valid_moves(self): # valid chess valid_moves
+
+		"""
+			this function updates the get_possible_moves to valid chess valid_moves
+			it identifies if the king is in check and provides solution to avoid checkmate or stalemate
+			it returns the updated possible valid moves and current_turn
+
+			it calls the get_possible_moves function and incheck function
+		"""
+		moves, turn = self.get_possible_moves() # passing in two varibles moves and turns because it returns 2
+		for i in range(len(moves)-1,-1,-1): # iterating the index of the moves from the last to 1st
+											# because we will be removing moves which will skip moves if iterated from 1st to last
+			self.make_move(moves[i]) 		#making a move to check if it will leads to the king being attacked
+			self.light_to_move = not  self.light_to_move # switching turn
+			if self.incheck():  # checking if opposition move attack king
+				moves.remove(moves[i]) # remove moves leading to king being attacked
+			self.light_to_move  = not self.light_to_move #switching back
+			self.undo_move()
+
+		return moves, turn
+
+
+
+	def incheck(self):
```

Comment:
use snake_case for method names for consistency with other methods. I.e. in_check not incheck